README.md
The idea for the dsEeprom-Library was to make my life easier. I did always the same job in several small projects on ESP-modules or for an Arduino: read/write EEPROM content.
To simplify this, I created this library. Because it is designed for my needs, it may or my not be useful for you.

General:
To install the library, you may download the zip-file or clone the repository.





CHANGELOG
dsEeprom - a library that provides extended 
         - access to the internal EEPROM
============================================
dsEeprom 0.9.1 / 10/29/2016
 * Initial upload
=========================================
dsEeprom 0.9.2 / 11/10/2016
 * fixed bug in restore raw
=========================================
dsEeprom 0.9.3 / 11/11/2016
 * added loggin initialisation
=========================================
dsEeprom 0.9.4 / 08/25/201z
 * removed implicit use of SimpleLog
=========================================


dsEeprom.h
//
// ************************************************************************
// dsEeprom
// (C) 2016 Dirk Schanz aka dreamshader
// ************************************************************************
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// ************************************************************************
//
//   A library to provide simplified access to the onchip EEPROM of an
//   Arduino or ESP8266.
//   This includes a checksum over the whole content, that is 
//   automatically stored and checked on EEPROM access, a mgic byte, 
//   that signals a valid content and provides information regarding 
//   the software version with that the EEPROM was written, simplified 
//   store and restore of strings, char array and so on.
//   Please refer to the specific funtion descrition for further
//   information.
//
// ************************************************************************
//
//
//-------- History --------------------------------------------------------
//
// 2016/10/28: initial version 
// 
//
// ************************************************************************

#ifndef _DSEEPROM_H_
#define _DSEEPROM_H_

#include <inttypes.h>
#include <stdarg.h>
#include <EEPROM.h>

#ifdef USE_SIMPLE_LOG
#include <SimpleLog.h>
#else
#define LOGLEVEL_QUIET    0
#define LOGLEVEL_DEFAULT  0
#define LOGLEVEL_INFO     0
#endif // USE_SIMPLE_LOG


#if defined(ARDUINO) && ARDUINO >= 100
	#include "Arduino.h"
#else
	#include "WProgram.h"
#endif

extern "C" {
}

//
//
// EEPROM_MAX_SIZE depending on MCU in use
// 
//   512 bytes on ATmega168 and ATmega8, 
//  1024 bytes on ATmega328  
//  4096 bytes on ATmega1280 and ATmega2560
//  up to 4096 bytes on ESP8266
//
#if defined (__AVR_ATmega8__) || defined (__AVR_ATmega168__)
#define EEPROM_MAX_SIZE                   512
#endif
//
#if defined (__AVR_ATmega328P__) || defined (__AVR_ATmega32U4__)
#define EEPROM_MAX_SIZE                  1024
#endif
//
#if defined (__AVR_ATmega1280__) || defined (__AVR_ATmega2560__)
#define EEPROM_MAX_SIZE                  4096
#endif
//
#ifdef ESP8266
#define EEPROM_MAX_SIZE                  4096
#endif // ESP8266
//
//
// minimal error codes
//
#define E_SUCCESS        0
#define E_BAD_CRC       -3
#define E_INVALID_MAGIC -2
//
// The member function version2Magic() has no real funtionality at this time.
// It simply returns the defined value of EEPROM_MAGIC_BYTE
// in a later version this may be replaced by a real calculation or
// mapping 
//
#define EEPROM_MAGIC_BYTE              0x7e
//
#define EEPROM_LEADING_LENGTH             2  // means two byte representing 
//                                           // the real length of the data field
#define EEPROM_MAXLEN_MAGIC               1
#define EEPROM_MAXLEN_CRC32               4
//
#define EEPROM_MAXLEN_BOOLEAN             1
#define EEPROM_MAXLEN_LONG                4
#define EEPROM_MAXLEN_SHORT               2
#define EEPROM_MAXLEN_CHAR                1
//
#define EEPROM_MAXLEN_WLAN_SSID          32  // max. length a SSID may have
#define EEPROM_MAXLEN_WLAN_PASSPHRASE    64  // max. length of a WLAN passphrase
#define EEPROM_MAXLEN_SERVER_IP          19  // max. length for the server IP
#define EEPROM_MAXLEN_SERVER_PORT         4  // max. length for the server port
#define EEPROM_MAXLEN_NODENAME           32  // max. lenght of the (generated) nodename
#define EEPROM_MAXLEN_ADMIN_PASSWORD     32  // max. length for admin password
//
//
// predefined standard layout of the eeprom:
//
#define EEPROM_HEADER_BEGIN         0
//
#define EEPROM_POS_MAGIC            0
//
#define EEPROM_POS_CRC32            (EEPROM_POS_MAGIC + EEPROM_MAXLEN_MAGIC)
//
#define EEPROM_HEADER_END           (EEPROM_POS_CRC32 + EEPROM_MAXLEN_CRC32 + EEPROM_LEADING_LENGTH)
//
// data area begins here
//
#define EEPROM_STD_DATA_BEGIN       EEPROM_HEADER_END       
//
#define EEPROM_POS_WLAN_SSID        EEPROM_STD_DATA_BEGIN
//
#define EEPROM_POS_WLAN_PASSPHRASE  (EEPROM_POS_WLAN_SSID + EEPROM_MAXLEN_WLAN_SSID + EEPROM_LEADING_LENGTH)
//
#define EEPROM_POS_SERVER_IP        (EEPROM_POS_WLAN_PASSPHRASE + EEPROM_MAXLEN_WLAN_PASSPHRASE + EEPROM_LEADING_LENGTH)
//
#define EEPROM_POS_SERVER_PORT      (EEPROM_POS_SERVER_IP + EEPROM_MAXLEN_SERVER_IP + EEPROM_LEADING_LENGTH)
//
#define EEPROM_POS_NODENAME         (EEPROM_POS_SERVER_PORT + EEPROM_MAXLEN_SERVER_PORT + EEPROM_LEADING_LENGTH)
//
#define EEPROM_POS_ADMIN_PASSWORD   (EEPROM_POS_NODENAME + EEPROM_MAXLEN_NODENAME + EEPROM_LEADING_LENGTH)
//
#define EEPROM_STD_DATA_END         (EEPROM_POS_ADMIN_PASSWORD + EEPROM_MAXLEN_ADMIN_PASSWORD + EEPROM_LEADING_LENGTH)
//
#define EEPROM_EXT_DATA_BEGIN       EEPROM_STD_DATA_END
//
// ----- the above region is reserved for standard values
//
// EEPROM status byte may be a combination of the following values
//
// ----- non failure/error indicator
//
#define EE_STATUS_OK_AND_READY   0
#define EE_STATUS_MODIFIED       1
#define EE_STATUS_COMMITED       2
//
// ----- failure/error indicators
//
#define EE_STATUS_INVALID_CRC    4
#define EE_STATUS_INVALID_MAGIC  8
#define EE_STATUS_INVALID_SIZE  16

// macro to check whether log output is done
//
#define DOLOG            (logLevel > LOGLEVEL_QUIET)

class dsEeprom {

  private:
    short logLevel;
    short status;
    unsigned char magic;
    int blockSize;
    unsigned int reSized;
    unsigned long crc32Old;
    unsigned long crc32New;

  public:
    dsEeprom( unsigned int blockSize = 0, unsigned char magic = 0x00, int logLevel = LOGLEVEL_QUIET );
    int init( unsigned int blockSize = 0, unsigned char magic = 0x00, int logLevel = LOGLEVEL_QUIET );
    virtual ~dsEeprom();
    short getStatus( void );
    void setBlocksize( unsigned int newSize );
    unsigned int getBlocksize( void );
    void setMagic( short newMagic );
    unsigned char getMagic( void );
    void setLoglevel( short newValue );
    short getLoglevel( void );
    unsigned char version2Magic( void );
    unsigned long crc( int startPos, int length );
    void wipe( void );
    int storeFieldLength( char* len, int dataIndex );
    int restoreFieldLength( char* len, int dataIndex );
    int storeBoolean(  char* data, int dataIndex );
    int restoreBoolean( char *data, int dataIndex );
    int storeRaw( const char* data, short len, int dataIndex );
    int restoreRaw( char* data, int dataIndex, int len, int maxLen);
    int storeBytes( const char* data, short len, int dataIndex );
    int restoreBytes( String& data, int dataIndex, int len, int maxLen);
    int storeString( String data, int maxLen, int dataIndex );
    int restoreString( String& data, int dataIndex, int maxLen );
    bool isValid();
    bool validate();
};



#endif // _DSEEPROM_H_


dsEeprom.cpp
//
// ************************************************************************
// dsEeprom
// (C) 2016 Dirk Schanz aka dreamshader
// ************************************************************************
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// ************************************************************************
//
//   A library to provide simplified access to the onchip EEPROM of an
//   Arduino or ESP8266.
//   This includes a checksum over the whole content, that is 
//   automatically stored and checked on EEPROM access, a mgic byte, 
//   that signals a valid content and provides information regarding 
//   the software version with that the EEPROM was written, simplified 
//   store and restore of strings, char array and so on.
//   Please refer to the specific funtion descrition for further
//   information.
//   I'll add additional dataformats now and then. Keep an eye to my
//   github repo ...
//
// ************************************************************************
//
//
//-------- History --------------------------------------------------------
//
// 2016/10/28: initial version 
// 
//
// ************************************************************************
//
#undef USE_SIMPLE_LOG


#include <Arduino.h>
#include <dsEeprom.h>

#ifdef USE_SIMPLE_LOG
#include <SimpleLog.h>
#endif // USE_SIMPLE_LOG



// ************************************************************************
// CRC lookup table
// ************************************************************************
//
static const PROGMEM uint32_t crc_table[16] = {
    0x00000000, 0x1db71064, 0x3b6e20c8, 0x26d930ac,
    0x76dc4190, 0x6b6b51f4, 0x4db26158, 0x5005713c,
    0xedb88320, 0xf00f9344, 0xd6d6a3e8, 0xcb61b38c,
    0x9b64c2b0, 0x86d3d2d4, 0xa00ae278, 0xbdbdf21c
};

#ifdef USE_SIMPLE_LOG
// ************************************************************************
// logger for debug/control output
// ************************************************************************
//
static SimpleLog Logger;
#endif // USE_SIMPLE_LOG

//
// ************************************************************************
// CRC calculation e.g. over EEPROMo for verification
// ************************************************************************
//
unsigned long dsEeprom::crc( int startPos, int length )
{
  unsigned long crc = ~0L;

  for (int index = startPos; index < (startPos + length); ++index) 
  {
    crc = crc_table[(crc ^ EEPROM.read(index)) & 0x0f] ^ (crc >> 4);
    crc = crc_table[((crc ^ EEPROM.read(index)) >> 4) & 0x0f] ^ (crc >> 4);
    crc = ~crc;
  }

#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
  if( DOLOG )
  {
    Logger.Log(LOGLEVEL_DEBUG, (const char*) "eeprom in crc: new value is %x\n", crc);
  }
#endif // DEBUG
#endif // USE_SIMPLE_LOG

  return crc;
}
//
// ************************************************************************
// EEPROM access
// ************************************************************************
//

dsEeprom::dsEeprom( unsigned int newBlockSize, unsigned char newMagic, int newLogLevel )
{

  status = 0;

  if( newLogLevel < LOGLEVEL_QUIET || newLogLevel > LOGLEVEL_INFO )
  {
    logLevel = LOGLEVEL_DEFAULT;
  }
  else
  {
    logLevel = newLogLevel;
  }

#ifdef USE_SIMPLE_LOG
  Logger.Init(logLevel, &Serial);
#endif // USE_SIMPLE_LOG


  if( newBlockSize <= 0 || newBlockSize > EEPROM_MAX_SIZE )
  {
    status |= EE_STATUS_INVALID_SIZE;
  }
  else
  {
#ifdef ESP8266
    EEPROM.begin(newBlockSize);
#else
    EEPROM.begin();
#endif // ESP8266
    blockSize = newBlockSize;
    status &= ~EE_STATUS_INVALID_SIZE;
  }

  if( magic == 0x00 )
  {
    status |= EE_STATUS_INVALID_MAGIC;
  }
  else
  {
    magic = newMagic;
  }

}

dsEeprom::~dsEeprom()
{
}

//
// ************************************************************************
// init
// ************************************************************************
//

int dsEeprom::init( unsigned int newBlockSize, unsigned char newMagic, int newLogLevel )
{

  status = EE_STATUS_OK_AND_READY;

  if( newLogLevel < LOGLEVEL_QUIET || newLogLevel > LOGLEVEL_INFO )
  {
    logLevel = LOGLEVEL_DEFAULT;
  }
  else
  {
    logLevel = newLogLevel;
  }

#ifdef USE_SIMPLE_LOG
  Logger.Init(logLevel, &Serial);
#endif // USE_SIMPLE_LOG

  if( newBlockSize <= 0 || newBlockSize > EEPROM_MAX_SIZE )
  {
    status |= EE_STATUS_INVALID_SIZE;
  }
  else
  {
#ifdef ESP8266
    EEPROM.begin(newBlockSize);
#else
    EEPROM.begin();
#endif // ESP8266
    blockSize = newBlockSize;
    status &= ~EE_STATUS_INVALID_SIZE;
  }

  if( newMagic == 0x00 )
  {
    status |= EE_STATUS_INVALID_MAGIC;
  }
  else
  {
    magic = newMagic;
  }

  return( status );
}


//
// tell status of dsEeprom-instance
//
short dsEeprom::getStatus( void )
{
  return status;
}

//
// clear EEPROM content (set to zero )
//
void dsEeprom::wipe( void )
{
  if( blockSize > 0 && blockSize <= EEPROM_MAX_SIZE )
  {
    for( int index = 0; index < blockSize; index++ )
    {
      EEPROM.write( index, '\0' );
    }

#ifdef ESP8266
    EEPROM.commit();
#endif // ESP8266
 
  }
  else
  {
    status |= EE_STATUS_INVALID_SIZE;
  }

}
//
//
//
// store field length to a specific position
//
int dsEeprom::storeFieldLength( char* len, int dataIndex )
{
  int retVal = 0;


  if( status & EE_STATUS_INVALID_SIZE )
  {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "eeprom has status EE_STATUS_INVALID_SIZE\n");
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG

  }
  else
  {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "write LEN byte [%x] to pos %d\n", len[0], dataIndex);
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
  
    EEPROM.write(dataIndex, len[0]);

#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "write LEN byte [%x] to pos %d\n", len[1], dataIndex+1);
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
  
    EEPROM.write(dataIndex+1, len[1]);

  }

  return(retVal);
}

//
// restore field length from a specific position
//
int dsEeprom::restoreFieldLength( char* len, int dataIndex )
{
  int retVal = 0;

  if( status & EE_STATUS_INVALID_SIZE )
  {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "eeprom has status EE_STATUS_INVALID_SIZE\n");
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
  }
  else
  {
    len[0] = EEPROM.read(dataIndex);

#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "got LEN byte [%x] from pos %d\n", len[0], dataIndex);
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG

    len[1] = EEPROM.read(dataIndex+1);

#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "got LEN byte [%x] from pos %d\n", len[1], dataIndex+1);
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
  }

  return(retVal);
}

//
// store a boolean value to a specific position
//
int dsEeprom::storeBoolean(  char* data, int dataIndex )
{
    int retVal = 0;
    short len = EEPROM_MAXLEN_BOOLEAN;

    if( status & EE_STATUS_INVALID_SIZE )
    {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
        if( DOLOG )
        {
            Logger.Log(LOGLEVEL_DEBUG, (const char*) "eeprom has status EE_STATUS_INVALID_SIZE\n");
        }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
    }
    else
    {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
        if( DOLOG )
        {
            Logger.Log(LOGLEVEL_DEBUG, (const char*) "store boolean to eeprom: Address is [%d]\n", dataIndex);
        }
#endif // DEBUG
#endif // USE_SIMPLE_LOG

        if( (retVal = storeFieldLength( (char*) &len, dataIndex )) == 0 )
        {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
            if( DOLOG )
            {
                Logger.Log(LOGLEVEL_DEBUG, (const char*) "Wrote:");
            }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
    
            for (int i = 0; i < len; ++i)
            {
                EEPROM.write(dataIndex + EEPROM_LEADING_LENGTH + i, data[i]);
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
                if( DOLOG )
                {
                    Logger.Log(LOGLEVEL_DEBUG, (const char*) " %x", data[i]); 
                }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
            }
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
            if( DOLOG )
            {
                Logger.Log(LOGLEVEL_DEBUG, (const char*) "\n");
            }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
        }
    }

    return(retVal);
}
//
// restore a boolean value from a specific position
//
int dsEeprom::restoreBoolean( char *data, int dataIndex )
{
    int retVal = 0;
    char rdValue;

    if( status & EE_STATUS_INVALID_SIZE )
    {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
      if( DOLOG )
      {
          Logger.Log(LOGLEVEL_DEBUG, (const char*) "eeprom has status EE_STATUS_INVALID_SIZE\n");
      }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
    }
    else
    {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
        if( DOLOG )
        {
            Logger.Log(LOGLEVEL_DEBUG, (const char*) "restore boolean from eeprom: Address is [%d]\n", dataIndex);
        }
#endif // DEBUG
#endif // USE_SIMPLE_LOG

        rdValue = EEPROM.read(dataIndex+ EEPROM_LEADING_LENGTH);

        if( rdValue == 0 )
        {
            data[0] = false;
        }
        else
        {
            if( rdValue == 1 )
            {
                data[0] = true;
            }
            else
            {
                retVal = -1;
            }
        }
    }

    return(retVal);
}
//
// store a raw byte buffer without leading length field
// to a specific position
//
int dsEeprom::storeRaw( const char* data, short len, int dataIndex )
{
    int retVal = 0;

    if( status & EE_STATUS_INVALID_SIZE )
    {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
      if( DOLOG )
      {
          Logger.Log(LOGLEVEL_DEBUG, (const char*) "eeprom has status EE_STATUS_INVALID_SIZE\n");
      }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
    }
    else
    {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
        if( DOLOG )
        {
            Logger.Log(LOGLEVEL_DEBUG, (const char*) "store raw data to eeprom: Address is [%d] - len = %d\n",
                dataIndex, len );
        }
#endif // DEBUG
#endif // USE_SIMPLE_LOG

#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
        if( DOLOG )
        {
            Logger.Log(LOGLEVEL_DEBUG, (const char*) "Wrote:");
        }
#endif // DEBUG
#endif // USE_SIMPLE_LOG

        for (int i = 0; i < len; ++i)
        {
            EEPROM.write(dataIndex+i, data[i]);

#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
            if( DOLOG )
            {
                Logger.Log(LOGLEVEL_DEBUG, (const char*) " wr[%d] -> %x\n", i, data[i]); 
            }

#endif // DEBUG
#endif // USE_SIMPLE_LOG
        }
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
        if( DOLOG )
        {
            Logger.Log(LOGLEVEL_DEBUG, (const char*) "\n");
        }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
    }

    return(retVal);

}
//
// restore a byte buffer without leading length field
// from a specific position 
//
int dsEeprom::restoreRaw( char* data, int dataIndex, int len, int maxLen)
{
  int retVal = 0;
  char c;
  
  if( status & EE_STATUS_INVALID_SIZE )
  {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "eeprom has status EE_STATUS_INVALID_SIZE\n");
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
  }
  else
  {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "restore raw data from eeprom: Address is [%d] - maxlen = %d\n",
                dataIndex, maxLen);
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG

    if( len > 0 )
    {
      for( int i=0; i < len && i < maxLen; i++ )
      {
        c = EEPROM.read(dataIndex + i);
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
        if( DOLOG )
        {
          Logger.Log(LOGLEVEL_DEBUG, (const char*) "rd[%d] <- %x\n", i, c);
        }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
        *data++ = c;
      }
    }

#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) " - done!");
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
  }

  return(retVal);
}

//
// store a byte array to a specific position
//
int dsEeprom::storeBytes( const char* data, short len, int dataIndex )
{
    int retVal = 0;

    if( status & EE_STATUS_INVALID_SIZE )
    {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
      if( DOLOG )
      {
        Logger.Log(LOGLEVEL_DEBUG, (const char*) "eeprom has status EE_STATUS_INVALID_SIZE\n");
      }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
    }
    else
    {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
        if( DOLOG )
        {
            Logger.Log(LOGLEVEL_DEBUG, (const char*) "store bytes to eeprom: Address is [%d] - len = %d\n",
                dataIndex, len );
        }
#endif // DEBUG
#endif // USE_SIMPLE_LOG

        if( (retVal = storeFieldLength( (char*) &len, dataIndex )) == 0 )
        {

#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
            if( DOLOG )
            {
                Logger.Log(LOGLEVEL_DEBUG, (const char*) "Wrote:\n");
            }
#endif // DEBUG
#endif // USE_SIMPLE_LOG

            for (int i = 0; i < len; ++i)
            {
                EEPROM.write(dataIndex + EEPROM_LEADING_LENGTH + i, data[i]);

#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
                if( DOLOG )
                {
                    Logger.Log(LOGLEVEL_DEBUG, (const char*) " wr -> %x\n", data[i]); 
                }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
            }
        }
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
        if( DOLOG )
        {
            Logger.Log(LOGLEVEL_DEBUG, (const char*) "\n");
        }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
    }

    return(retVal);
}


//
// restore a string var from a specific position
//
int dsEeprom::restoreBytes( String& data, int dataIndex, int len, int maxLen)
{
  int retVal = 0;
  char c;
  
  if( status & EE_STATUS_INVALID_SIZE )
  {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "eeprom has status EE_STATUS_INVALID_SIZE\n");
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
  }
  else
  {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "restore data from eeprom: Address is [%d] - maxlen = %d\n",
                dataIndex, maxLen);
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG

    if( len > 0 )
    {
      data = "";
      for( int i=0; i < len && i < maxLen; i++ )
      {
        c = EEPROM.read(dataIndex + EEPROM_LEADING_LENGTH + i);
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
        if( DOLOG )
        {
          Logger.Log(LOGLEVEL_DEBUG, (const char*) "rd <- %c", c);
        }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
        data += c;
      }
    }

#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) " - done!\n");
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
  }

  return(retVal);
}

//
// store a string var to a specific position
//
int dsEeprom::storeString( String data, int maxLen, int dataIndex )
{
  int retVal = 0;
  short wrLen;

  if( status & EE_STATUS_INVALID_SIZE )
  {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "eeprom has status EE_STATUS_INVALID_SIZE\n");
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
  }
  else
  {
    data.trim();

    if( data.length() <= (unsigned int) maxLen )
    {
      wrLen = data.length();
    }
    else
    {
      wrLen = maxLen;
    }

    retVal = storeBytes( data.c_str(), wrLen, dataIndex );
  }

  return(retVal);
}
//
// restore a string var from a specific position
//
int dsEeprom::restoreString( String& data, int dataIndex, int maxLen )
{
  int retVal = 0;
  short len = 0;
  
  if( status & EE_STATUS_INVALID_SIZE )
  {
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "eeprom has status EE_STATUS_INVALID_SIZE\n");
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
  }
  else
  {
    if( (retVal = restoreFieldLength( (char*) &len, dataIndex )) == 0 )
    {
      retVal = restoreBytes( data, dataIndex, len, maxLen );
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
      if( DOLOG )
      {
        Logger.Log(LOGLEVEL_DEBUG, (const char*) " - done!\n");
      }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
    }
  }

  return(retVal);
}

//
// check whether first byte in EEPROM is "magic"
//
bool dsEeprom::isValid()
{
  bool retVal = true;
  unsigned char rdMagic;

  if( magic == 0 || (rdMagic = EEPROM.read( EEPROM_POS_MAGIC )) !=  magic )
  {
    retVal = false;
#ifdef USE_SIMPLE_LOG
#ifdef DEBUG
    if( DOLOG )
    {
      Logger.Log(LOGLEVEL_DEBUG, (const char*) "wrong magic: %x should be %x\n", rdMagic, magic);
    }
#endif // DEBUG
#endif // USE_SIMPLE_LOG
  }
  else
  {
    magic = rdMagic;
  }

  return(retVal);
}

//
// place a "magic" to the first byte in EEPROM
//
bool dsEeprom::validate()
{
    bool retVal = true;

    if( blockSize > 0 && blockSize <= EEPROM_MAX_SIZE )
    {
        EEPROM.write( EEPROM_POS_MAGIC, magic );
        this->crc32Old = crc( EEPROM_STD_DATA_BEGIN, this->blockSize );
        this->crc32New = this->crc32Old;
        storeRaw( (char*) &this->crc32Old, EEPROM_MAXLEN_CRC32, EEPROM_POS_CRC32 );

#ifdef ESP8266
    EEPROM.commit();
#endif // ESP8266

    }
    else
    {
        status |= EE_STATUS_INVALID_SIZE;
    }

    return(retVal);
}

void dsEeprom::setBlocksize( unsigned int newSize )
{
  if( newSize > 0 && newSize <= EEPROM_MAX_SIZE )
  {
    blockSize = newSize;
  }
  else
  {
    status |= EE_STATUS_INVALID_SIZE;
  }
}

//
//
//
unsigned int dsEeprom::getBlocksize( void )
{
    return( blockSize );
}

//
//
//
void dsEeprom::setMagic( short newMagic )
{
  if( newMagic != 0x00 )
  {
    magic = newMagic;
  }
}

//
//
//
unsigned char dsEeprom::getMagic( void )
{
  return( magic );
}

//
//
//
void dsEeprom::setLoglevel( short newValue )
{
  if( newValue < LOGLEVEL_QUIET || newValue > LOGLEVEL_INFO )
  {
    logLevel = LOGLEVEL_DEFAULT;
  }
  else
  {
    logLevel = newValue;
  }
}

//
//
//
short dsEeprom::getLoglevel( void )
{
  return( logLevel );
}

//
//
//

unsigned char dsEeprom::version2Magic( void )
{
  return( EEPROM_MAGIC_BYTE );
}




